src/contracts/atlas/GasAccounting.sol

```diff
@@ -201,107 +208,187 @@ abstract contract GasAccounting is SafetyLocks {
+        if (ctx.solverCount > 0) {
+            // Calculate the unadjusted bundler gas surcharge
+            uint256 _grossBundlerGasSurcharge =
+                (_claims * BUNDLER_SURCHARGE_RATE) / (SURCHARGE_SCALE + ATLAS_SURCHARGE_RATE + BUNDLER_SURCHARGE_RATE);
+
+            // Calculate an estimate for how much gas should be remaining
+            // NOTE: There is a free buffer of one SolverOperation because solverIndex starts at 0.
+            uint256 _upperGasRemainingEstimate =
+                (solverGasLimit * (ctx.solverCount - ctx.solverIndex)) + _BUNDLER_GAS_PENALTY_BUFFER;
+
+            // Increase the _writeoffs value if the bundler set too high of a gas parameter and forced solvers to
+            // maintain higher escrow balances.
+            if (_gasLeft > _upperGasRemainingEstimate) {
+                // Penalize the bundler's gas
+                uint256 _bundlerGasOveragePenalty =
+                    _grossBundlerGasSurcharge - (_grossBundlerGasSurcharge * _upperGasRemainingEstimate / _gasLeft);
+                _writeoffs += _bundlerGasOveragePenalty;
+            }
+        }
     }
 
     /// @notice Settles the transaction after execution, determining the final distribution of funds between the winning
     /// solver and the bundler based on the outcome.
     /// @dev This function adjusts the claims, withdrawals, deposits, and surcharges based on the gas used by the
     /// transaction.
     /// @param ctx Context struct containing relavent context information for the Atlas auction.
-    /// @param winningSolver The address of the winning solver.
     function _settle(
         Context memory ctx,
-        address winningSolver
+        uint256 solverGasLimit
     )
         internal
-        returns (uint256 claimsPaidToBundler, uint256 netGasSurcharge)
+        returns (uint256 claimsPaidToBundler, uint256 netAtlasGasSurcharge)
     {
-        // NOTE: If there is no winningSolver but the dApp config allows unfulfilled 'successes,' the bundler
-        // is treated as the Solver.
+        // NOTE: If there is no winning solver but the dApp config allows unfulfilled 'successes', the bundler
+        // is treated as the solver.
+
+        // If a solver won, their address is still in the _solverLock
+        address _winningSolver = address(uint160(_solverLock));
 
         // Load what we can from storage so that it shows up in the gasleft() calc
         uint256 _surcharge = cumulativeSurcharge;
-        uint256 _claims = claims;
         uint256 _withdrawals = withdrawals;
         uint256 _deposits = deposits;
+        (uint256 _claims, uint256 _writeoffs) = _getAdjustedClaimsAndWriteoffs(ctx, solverGasLimit);
+
+        netAtlasGasSurcharge =
+            (_claims * ATLAS_SURCHARGE_RATE) / (SURCHARGE_SCALE + ATLAS_SURCHARGE_RATE + BUNDLER_SURCHARGE_RATE);
+
+        // Update the stored cumulative surcharge
+        cumulativeSurcharge = _surcharge + netAtlasGasSurcharge;
```

Comment:
We optimistically add `netAtlasGasSurcharge` to `cumulativeSurcharge`, I think with the intention of reverting in a case where we dont get to keep the surcharge. But there are paths below where we don't keep it and it doesn't revert. In these cases there would be an accounting imbalance.

E.g. if no solver successful AND `_deposits >= _withdrawals`

Auditors recommend always charging the surcharge to prevent spam: https://github.com/spearbit-audits/review-fastlane/issues/55

Reply:
See discussion in comment below. This is fixed in the latest changes.